video/out/vo_sixel.c

```diff
@@ -0,0 +1,429 @@
+        if (priv->dither == NULL)
+            return SIXEL_FALSE;
+        sixel_dither_set_diffusion_type(priv->dither, priv->opts->diffuse);
+    }
+    return SIXEL_OK;
+}
+
+static SIXELSTATUS prepare_dynamic_palette(struct vo *vo)
+{
+    SIXELSTATUS status = SIXEL_FALSE;
+    struct priv *priv = vo->priv;
+
+    /* create histgram and construct color palette
+     * with median cut algorithm. */
+    status = sixel_dither_initialize(priv->testdither, priv->buffer,
+                                     priv->opts->width, priv->opts->height, 3,
+                                     LARGE_NORM, REP_CENTER_BOX,
+                                     QUALITY_LOW);
+    if (SIXEL_FAILED(status))
+        return status;
+
+    if (detect_scene_change(vo)) {
+        if (priv->dither)
+            sixel_dither_unref(priv->dither);
+
+        priv->dither = priv->testdither;
+        status = sixel_dither_new(&priv->testdither, priv->opts->reqcolors, NULL);
+
+        if (SIXEL_FAILED(status))
+            return status;
+
+        sixel_dither_set_diffusion_type(priv->dither, priv->opts->diffuse);
+    } else
+        sixel_dither_set_body_only(priv->dither, 1);
+
+    return status;
+}
+
+static int resize(struct vo *vo)
+{
+    struct priv *priv = vo->priv;
+
+    talloc_free(priv->buffer);
+
+    if (priv->dither) {
+        sixel_dither_unref(priv->dither);
+        priv->dither = NULL;
+    }
+
+    if (priv->testdither) {
+        sixel_dither_unref(priv->testdither);
+        priv->testdither = NULL;
+    }
+
+    SIXELSTATUS status = sixel_dither_new(&priv->testdither, priv->opts->reqcolors, NULL);
+
+    if (SIXEL_FAILED(status))
+        return status;
+
+    priv->buffer =
+        talloc_array(NULL, uint8_t, depth * priv->opts->width * priv->opts->height);
+
+    return 0;
+}
+
+static int reconfig(struct vo *vo, struct mp_image_params *params)
+{
+    struct priv *priv = vo->priv;
+    priv->image_height = params->h;
+    priv->image_width  = params->w;
+    priv->image_format = params->imgfmt;
+    save_image_offset_location(vo);
+
+    priv->sws->src = *params;
+    priv->sws->dst = (struct mp_image_params) {
+        .imgfmt = IMGFMT,
+        .w = priv->opts->width,
+        .h = priv->opts->height,
+        .p_w = 1,
+        .p_h = 1,
+    };
+
+    priv->frame = mp_image_alloc(IMGFMT, priv->opts->width, priv->opts->height);
+    if (!priv->frame)
+        return -1;
+
+    if (mp_sws_reinit(priv->sws) < 0)
+        return -1;
+
+    printf(ESC_HIDE_CURSOR);
+    printf(ESC_CLEAR_SCREEN);
+    vo->want_redraw = true;
+
+    return resize(vo);
+}
+
+static void draw_image(struct vo *vo, mp_image_t *mpi)
+{
+    struct priv *priv = vo->priv;
+    struct mp_image src = *mpi;
+
+    // Downscale the image
+    mp_sws_scale(priv->sws, priv->frame, &src);
+
+    // Copy from mpv to RGB format as required by libsixel
+    memcpy_pic(priv->buffer, priv->frame->planes[0], priv->opts->width * depth, priv->opts->height,
```

Comment:
Is there an mpv image format which is already suitable for libsixel without this additional conversion?

Reply:
I'll ask in the mpv-devel group since I don't have extensive knowledge about mpv APIs.

Reply:
OK, let's leave it like this for now. We might be able to improve it later.